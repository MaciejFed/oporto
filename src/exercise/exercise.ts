import { Comparable } from '../common/common';
import { RatioRange } from '../service/progress';
export type ExerciseType =
  | 'VerbExercise'
  | 'NounTranslation'
  | 'OtherTranslation'
  | 'AdjectiveTranslation'
  | 'VerbTranslation'
  | 'SentenceTranslation'
  | 'FitInGap';

export const translationTypes: ExerciseType[] = [
  'NounTranslation',
  'VerbTranslation',
  'OtherTranslation',
  'AdjectiveTranslation',
  'SentenceTranslation'
];

export interface Exercise extends Comparable {
  exerciseType: ExerciseType;
  getExerciseBodyPrefix(): string;
  getExerciseBodySuffix(): string;
  getExerciseDescription(): string;
  getExerciseTranslation(): string | undefined;
  getCorrectAnswer(): string;
  checkAnswerCorrect(answer: string): boolean;
  getRepeatAnswerPhrase(): string;
  getMaxWantedProgress(): RatioRange | '100+';
  getMinimumAnswers(): number;
}
