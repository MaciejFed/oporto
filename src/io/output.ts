import { terminal } from 'terminal-kit';
import clear from 'clear';
import { AppEventListener } from '../event/event-listener';
import { createColorArray } from './terminal/terminal-utils';

export enum Color {
  B = 'B',
  R = 'R',
  G = 'G',
  W = 'W',
  Y = 'Y'
}

export class ColoredText {
  public text: string;
  public colors: Color[];

  public constructor(text = '', colors: Color[] = []) {
    // eslint-disable-next-line no-param-reassign,no-unused-expressions
    colors.length === 0 ? (colors = createColorArray({ white: text.length })) : colors;
    if (text.length > colors.length) throw new Error('colors array lenght not equal to text lenght');
    this.text = text;
    this.colors = colors;
  }

  concat(coloredText: ColoredText): ColoredText {
    this.text = this.text.concat(coloredText.text);
    this.colors = this.colors.concat(coloredText.colors);
    return this;
  }

  padEnd(pad: number): ColoredText {
    if (pad < this.text.length) return this;
    const diff = pad - this.text.length;
    this.text = this.text.padEnd(pad);
    this.colors = this.colors.concat(Array(diff).fill(Color.W));

    return this;
  }
}

export class Output implements AppEventListener {
  private outputTable: string[][] = [];
  private colorTable: string[][] = [];
  private textColor: Color;
  private colorMap: Record<Color, () => Output> = {
    B: this.blue,
    R: this.red,
    G: this.green,
    W: this.white,
    Y: this.yellow
  };

  public constructor() {
    this.resetTable();
    this.textColor = Color.W;
  }

  public moveCursor(x: number, y: number) {
    terminal.moveTo(x, y);
  }

  public moveToColored(x: number, y: number, coloredText: ColoredText) {
    coloredText.text.split('').forEach((char, index) => {
      this.colorMap[coloredText.colors[index]]();
      this.moveTo(x + index, y, char);
    });
    this.white();
  }

  public moveToColoredRows(x: number, y: number, coloredText: ColoredText[]) {
    coloredText.forEach((line, index) => {
      this.moveToColored(x, y + index, line);
    });
  }

  public moveTo(x: number, y: number, text: string | undefined) {
    if (text) {
      if (text.includes('\n')) {
        text.split('\n').forEach((_v, index) => {
          this.outputTable[y + index][x] = 'IMAGE';
        });
      } else {
        for (let i = 0; i < text.length; i++) {
          if (text[i] === '\n') {
            // eslint-disable-next-line no-param-reassign
            y++;
            // eslint-disable-next-line no-param-reassign
            text = text?.slice(i + 1);
            // eslint-disable-next-line no-param-reassign
            i = 0;
          }
          this.outputTable[y][x + i] = text[i];
          this.colorTable[y][x + i] = this.textColor;
        }
      }
      terminal.moveTo(x, y, text);
    }
  }

  public yellow(): Output {
    this.textColor = Color.Y;
    terminal.yellow();
    return this;
  }

  public blue(): Output {
    this.textColor = Color.B;
    terminal.blue();
    return this;
  }

  public red(): Output {
    this.textColor = Color.R;
    terminal.red();
    return this;
  }

  public green(): Output {
    this.textColor = Color.G;
    terminal.green();
    return this;
  }

  public white(): Output {
    this.textColor = Color.W;
    terminal.white();
    return this;
  }

  public bold(isBold = true): Output {
    terminal.bold(isBold);
    return this;
  }

  public hideCursor(hideCursor = true): Output {
    terminal.hideCursor(hideCursor);
    return this;
  }

  public getOutput(): string {
    return this.outputTable.map((row) => row.join('')).join('\n');
  }

  public getColorOutput(): string {
    return this.colorTable.map((row) => row.join('')).join('\n');
  }

  public clearTerminal(): void {
    clear();
    this.resetTable();
  }

  private resetTable(): void {
    this.outputTable = new Array(60).fill(' ').map(() => new Array(70).fill(' '));
    this.colorTable = new Array(60).fill(' ').map(() => new Array(70).fill(' '));
  }

  public registerListeners(): void {}
}
