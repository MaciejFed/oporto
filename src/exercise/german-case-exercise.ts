import { Comparable } from '../common/common';
import { BaseWordType, Exercise, ExerciseType } from './exercise';
import { GenderWord, GermanCase, GermanCaseWord, GermanGender } from '../repository/german-exercises-repository';

const safe = (word?: string) => (word ? `[${word}]` : '');

export class GermanCaseExercise implements Exercise, Comparable {
  public exerciseType: ExerciseType;
  public caseWord: GermanCaseWord;
  public germanCase: GermanCase;
  public gender?: GermanGender;

  public constructor(caseWord: GermanCaseWord, germanCase: GermanCase, gender?: GermanGender) {
    this.exerciseType = 'GermanCaseExercise';
    this.caseWord = caseWord;
    this.germanCase = germanCase;
    this.gender = gender;
  }

  static new(caseWord: GermanCaseWord, germanCase: GermanCase, gender?: GermanGender): GermanCaseExercise {
    return new GermanCaseExercise(caseWord, germanCase, gender);
  }

  getTranslation = () => undefined;

  getBodyPrefix = () => `[${this.germanCase}]${safe(this.gender)}: `;

  getBodySuffix = () => '';

  getDescription = () => `Base: ${this.caseWord.english}`;

  getCorrectAnswer = () => {
    const word = this.caseWord.german[this.germanCase] || '';
    return this.getFinalWord(word);
  };

  isAnswerCorrect(answer: string): boolean {
    const correctAnswer = this.getCorrectAnswer();
    return correctAnswer.toLowerCase() === answer.toLowerCase();
  }

  getMinAnswerCount(): number {
    return 2;
  }

  getBaseWord() {
    return this.caseWord;
  }

  getBaseWordType(): BaseWordType {
    return BaseWordType.OTHER;
  }

  getBaseWordAsString(): string | undefined {
    const word = this.getBaseWord().german.nominative!;
    return this.getFinalWord(word, GermanGender.maskulinum);
  }

  getRetryPrompt = () => this.getCorrectAnswer();

  equal = (exercise: GermanCaseExercise) =>
    exercise.exerciseType === 'GermanCaseExercise' &&
    this.caseWord.english === exercise.caseWord.english &&
    this.gender === exercise.gender &&
    this.germanCase === exercise.germanCase;

  public getFinalWord(word: GenderWord | string, gender: GermanGender = GermanGender.maskulinum): string {
    if (typeof word === 'string') return word;
    const result = word[gender ?? this.gender!];
    if (!result) throw new Error(`Unexpected no word for [${JSON.stringify(this.caseWord, null, 4)}]`);
    return result;
  }
}
