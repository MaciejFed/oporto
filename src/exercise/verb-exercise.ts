import { Comparable } from '../common/common';
import { Person, Verb } from '../repository/exercises-repository';
import { getCorrectVerbConjugation, getRandomPerson, getRandomVerb } from '../service/verb';
import { Exercise, ExerciseType } from './exercise';
import { RatioRange } from '../service/progress';

export type VerbTime = 'presentSimple' | 'pastPerfect';

const vertTimePrintVersion = (verbTime: VerbTime) => {
  if (verbTime === 'presentSimple') return 'Present Simple';
  return 'Past Perfect';
};

export class VerbExercise implements Exercise, Comparable {
  exerciseType: ExerciseType;
  verbTime: VerbTime;
  verb: Verb;
  person: Person;

  constructor() {
    this.exerciseType = 'VerbExercise';
    this.verb = getRandomVerb();
    this.person = getRandomPerson();
    this.verbTime = 'presentSimple';
  }

  static new(verb: Verb, person: Person, verbTime: VerbTime): VerbExercise {
    const verbExercise = new VerbExercise();
    verbExercise.verb = verb;
    verbExercise.person = person;
    verbExercise.verbTime = verbTime;

    return verbExercise;
  }

  getExerciseTranslation = () => undefined;

  getExerciseBodyPrefix = () => `[${vertTimePrintVersion(this.verbTime)}] ${this.person}: `;

  getExerciseBodySuffix = () => '';

  getExerciseDescription = () => `Infinitive: ${this.verb.english}`;

  getCorrectAnswer = () => getCorrectVerbConjugation(this.verb, this.person, this.verbTime);

  checkAnswerCorrect(answer: string): boolean {
    const correctConjugation = this.getCorrectAnswer();
    return correctConjugation.toLowerCase() === answer.toLowerCase();
  }

  getMaxWantedProgress(): RatioRange | '100+' {
    return '80-100';
  }

  getMinimumAnswers(): number {
    return 2;
  }

  getRepeatAnswerPhrase = () => `${this.person} ${this.getCorrectAnswer()}`;

  equal = (exercise: VerbExercise) =>
    exercise.exerciseType === 'VerbExercise' &&
    this.verb.infinitive === exercise.verb.infinitive &&
    this.verbTime === exercise.verbTime &&
    this.person === exercise.person;
}
