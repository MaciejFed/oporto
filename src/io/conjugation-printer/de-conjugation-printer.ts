import { Result } from '../../service/result';
import { ColoredText } from '../output';
import { Exercise } from '../../exercise/exercise';
import { VerbConjugation } from './conjugation-printer';
import { GermanPerson, GermanVerb } from '../../repository/german-exercises-repository';
import { GermanVerbExercise } from '../../exercise/german-verb-exercise';
import { createColorArray } from '../terminal/terminal-utils';
import { Person } from '../../repository/exercises-repository';

type Conjugation = Record<GermanPerson, string>;

export class DEVerbConjugation extends VerbConjugation<GermanVerb> {
  public constructor(data: GermanVerb, result: Result[]) {
    super(data, result);
  }

  getPersonForX(x: number): GermanPerson {
    return Object.values(GermanPerson)[x] as GermanPerson;
  }

  checkStandardConjugation(x: number, y: number): ColoredText {
    const cellValue = this.getCell(x, y)!;
    const wordBase = this.data.infinitive.endsWith('en')
      ? this.data.infinitive.slice(0, -2)
      : this.data.infinitive.slice(0, -1);

    const person = this.getPersonForX(x);
    const standardConjugations: Conjugation = {
      Ich: 'e',
      Du: 'st',
      'Er/Sie/Es': 't',
      Wir: 'en',
      Ihr: 't',
      Sie: 'en'
    };
    const expectedVerbEnding = standardConjugations[person];
    const expectedVerbEndingLength = expectedVerbEnding.length;
    if (wordBase.concat(expectedVerbEnding) !== cellValue) {
      return new ColoredText(cellValue, createColorArray({ yellow: cellValue.length }));
    }
    return new ColoredText(
      cellValue,
      createColorArray({
        white: wordBase.length,
        green: expectedVerbEndingLength
      })
    );
  }

  getExercise(x: number, y: number): Exercise {
    // @ts-ignore
    return GermanVerbExercise.new(this.data, this.getPersonForX(x), this.getTenseForY(y));
  }

  getRowTitles(): string[] {
    return Object.keys(GermanPerson).map((person) => `${person}: `);
  }

  getTableSize(): { x: number; y: number } {
    return {
      y: 2,
      x: Object.values(this.data.presentSimple).length
    };
  }

  getRowTitle(x: number): ColoredText {
    return new ColoredText(`${Object.keys(GermanPerson)[x]}: `);
  }
}
