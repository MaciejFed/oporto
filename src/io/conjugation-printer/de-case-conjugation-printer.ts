import { Result } from '../../service/result';
import { ColoredText } from '../output';
import { Exercise } from '../../exercise/exercise';
import { VerbConjugation } from './conjugation-printer';
import {
  GermanCase,
  GermanCaseWord,
  GermanGender,
  GermanPerson,
  GermanVerb
} from '../../repository/german-exercises-repository';
import { GermanVerbExercise } from '../../exercise/german-verb-exercise';
import { createColorArray } from '../terminal/terminal-utils';
import { Person } from '../../repository/exercises-repository';
import { GermanCaseExercise } from '../../exercise/german-case-exercise';

export class DECaseConjugation extends VerbConjugation<GermanCaseWord> {
  public constructor(data: GermanCaseWord, result: Result[]) {
    super(data, result);
  }

  getCell(x: number, y: number): string | undefined {
    if (x > this.getTableSize().x || y > this.getTableSize().y) throw Error('');

    const valueTime = Object.values(this.data.german)[y];
    if (!valueTime) return undefined;
    if (typeof valueTime === 'string') return valueTime;
    // @ts-ignore
    return Object.values(valueTime)[x];
  }

  getPersonForX(x: number): GermanCase {
    return Object.values(GermanCase)[x] as GermanCase;
  }

  getGenderForY(y: number): GermanGender | undefined {
    if (this.isNonGender()) return undefined;
    return Object.values(GermanGender)[y] as GermanGender;
  }

  checkStandardConjugation(x: number, y: number): ColoredText {
    const cellValue = this.getCell(x, y)!;
    return new ColoredText(cellValue);
  }

  getExercise(x: number, y: number): Exercise {
    return GermanCaseExercise.new(this.data, this.getPersonForX(y), this.getGenderForY(x));
  }

  getRowTitles(): string[] {
    return Object.keys(GermanPerson).map((person) => `${person}: `);
  }

  getTableSize(): { x: number; y: number } {
    const valueTime = Object.values(this.data.german)[0];
    if (!valueTime || typeof valueTime === 'object') {
      return {
        x: Object.values(valueTime).length,
        y: Object.values(this.data.german).length
      };
    }
    return {
      x: 1,
      y: Object.values(this.data.german).length
    };
  }

  getRowTitle(x: number): ColoredText {
    if (this.isNonGender()) return new ColoredText('');
    return new ColoredText(`${Object.keys(this.data.german)[x]}: `);
  }

  isNonGender(): boolean {
    const valueTime = Object.values(this.data.german)[0];
    return !valueTime || typeof valueTime !== 'object';
  }
}
