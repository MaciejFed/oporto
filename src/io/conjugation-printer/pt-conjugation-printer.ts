import { Person, Verb } from '../../repository/exercises-repository';
import { Result } from '../../service/result';
import { ColoredText } from '../output';
import { VerbExercise, VerbTime } from '../../exercise/verb-exercise';
import { BaseWord, Exercise } from '../../exercise/exercise';
import { VerbConjugation } from './conjugation-printer';
import { createColorArray } from '../terminal/terminal-utils';
import { GermanPerson } from '../../repository/german-exercises-repository';

type VerbEnding = 'ar' | 'er' | 'ir';
const standardVerbEndings: VerbEnding[] = ['ar', 'er', 'ir'];

export const standardConjugations: Record<VerbTime, Record<VerbEnding, Record<Person, string>>> = {
  presentSimple: {
    ar: {
      [Person.Eu]: 'o',
      [Person.Tu]: 'as',
      [Person.ElaEleVocê]: 'a',
      [Person.Nós]: 'amos',
      [Person.ElesElasVosēs]: 'am'
    },
    er: {
      [Person.Eu]: 'o',
      [Person.Tu]: 'es',
      [Person.ElaEleVocê]: 'e',
      [Person.Nós]: 'emos',
      [Person.ElesElasVosēs]: 'em'
    },
    ir: {
      [Person.Eu]: 'o',
      [Person.Tu]: 'es',
      [Person.ElaEleVocê]: 'e',
      [Person.Nós]: 'imos',
      [Person.ElesElasVosēs]: 'em'
    }
  },
  pastPerfect: {
    ar: {
      [Person.Eu]: 'ei',
      [Person.Tu]: 'aste',
      [Person.ElaEleVocê]: 'ou',
      [Person.Nós]: 'ámos',
      [Person.ElesElasVosēs]: 'aram'
    },
    er: {
      [Person.Eu]: 'i',
      [Person.Tu]: 'este',
      [Person.ElaEleVocê]: 'eu',
      [Person.Nós]: 'emos',
      [Person.ElesElasVosēs]: 'eram'
    },
    ir: {
      [Person.Eu]: 'i',
      [Person.Tu]: 'iste',
      [Person.ElaEleVocê]: 'iu',
      [Person.Nós]: 'imos',
      [Person.ElesElasVosēs]: 'iram'
    }
  },
  imperfect: {
    ar: {
      [Person.Eu]: 'va',
      [Person.Tu]: 'vas',
      [Person.ElaEleVocê]: 'va',
      [Person.Nós]: 'vamos',
      [Person.ElesElasVosēs]: 'vam'
    },
    er: {
      [Person.Eu]: 'va',
      [Person.Tu]: 'vas',
      [Person.ElaEleVocê]: 'va',
      [Person.Nós]: 'vamos',
      [Person.ElesElasVosēs]: 'vam'
    },
    ir: {
      [Person.Eu]: 'va',
      [Person.Tu]: 'vas',
      [Person.ElaEleVocê]: 'va',
      [Person.Nós]: 'vamos',
      [Person.ElesElasVosēs]: 'vam'
    }
  }
};

export class PTVerbConjugation extends VerbConjugation<Verb> {
  public constructor(data: Verb, result: Result[]) {
    super(data, result);
  }

  getPersonForX(x: number): Person {
    return Object.values(Person)[x] as Person;
  }

  checkStandardConjugation(x: number, y: number): ColoredText {
    const cellValue = this.getCell(x, y)!;
    const wordBase = this.data.infinitive.slice(0, -2);

    const person = this.getPersonForX(x);
    const tense = this.getTenseForY(y);

    const normalize = (word: string) => {
      return (word.includes('-') ? word.substring(0, word.indexOf('-')) : word) as VerbEnding;
    };
    const verbEnding = normalize(this.data.infinitive).slice(-2) as VerbEnding;
    if (!standardVerbEndings.includes(verbEnding)) {
      return new ColoredText(cellValue, createColorArray({ yellow: cellValue.length }));
    }
    const expectedVerbEnding = standardConjugations[tense][verbEnding][person];
    const expectedVerbEndingLength = expectedVerbEnding.length;
    if (wordBase.concat(expectedVerbEnding) !== cellValue) {
      return new ColoredText(cellValue, createColorArray({ yellow: cellValue.length }));
    }
    return new ColoredText(
      cellValue,
      createColorArray({
        white: wordBase.length,
        green: expectedVerbEndingLength
      })
    );
  }

  getExercise(x: number, y: number): Exercise {
    return VerbExercise.new(this.data, this.getPersonForX(x), this.getTenseForY(y));
  }

  getRowTitles(): string[] {
    return Object.keys(Person).map((person) => `${person}: `);
  }

  getTableSize(): { x: number; y: number } {
    return {
      y: 3,
      x: Object.values(this.data.presentSimple).length
    };
  }

  getRowTitle(x: number): ColoredText {
    return new ColoredText(`${Object.keys(Person)[x]}: `);
  }
}
