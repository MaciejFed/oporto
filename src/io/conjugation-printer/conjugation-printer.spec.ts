import { Person, Verb, wordDatabase } from '../../repository/exercises-repository';
import { Color, ColoredText, Output } from '../output';
import { generateResultForExercise } from '../../priority/priority.util';
import { VerbExercise } from '../../exercise/verb-exercise';
import { PTVerbConjugation } from './pt-conjugation-printer';
import { GermanCase, GermanPerson, GermanVerb, readAllDE } from '../../repository/german-exercises-repository';
import { GermanVerbExercise } from '../../exercise/german-verb-exercise';
import { DEVerbConjugation } from './de-conjugation-printer';
import { DECaseConjugation } from './de-case-conjugation-printer';
import { GermanCaseExercise } from '../../exercise/german-case-exercise';

const generateResults = (verb: Verb) => [
  ...generateResultForExercise(VerbExercise.new(verb, Person.Eu, 'presentSimple'), true, 'keyboard', 1),
  ...generateResultForExercise(VerbExercise.new(verb, Person.Tu, 'presentSimple'), false, 'keyboard', 1)
];

const generateResultsDE = (verb: GermanVerb) => [
  ...generateResultForExercise(GermanVerbExercise.new(verb, GermanPerson.Ich, 'presentSimple'), true, 'keyboard', 1),
  ...generateResultForExercise(GermanVerbExercise.new(verb, GermanPerson.Du, 'presentSimple'), false, 'keyboard', 1)
];

describe('Conjugation Printer', () => {
  describe('Portuguese Verb', () => {
    it('renders standard', () => {
      const verb = wordDatabase.verb('conhecer');
      const results = generateResults(verb);
      const printer = new PTVerbConjugation(verb, results);

      const exception = printer.renderCell(0, 0);
      const nonException = printer.renderCell(1, 0);

      expect(exception).toEqual(
        new ColoredText('Eu:      conheço     🟢', Array(19).fill(Color.Y).concat(Array(6).fill(Color.W)))
      );
      expect(nonException).toEqual(
        new ColoredText(
          'conheces    ❌',
          Array(6).fill(Color.W).concat([Color.G, Color.G, Color.W, Color.W, Color.W, Color.W, Color.W])
        )
      );
    });

    it('renders exception', () => {
      const verb = wordDatabase.verb('pôr');
      const results = generateResults(verb);
      const printer = new PTVerbConjugation(verb, results);

      const exception = printer.renderCell(0, 0);
      const nonException = printer.renderCell(1, 0);

      expect(exception).toEqual(
        new ColoredText('ponho       ✅', Array(5).fill(Color.Y).concat(Array(8).fill(Color.W)))
      );
      expect(nonException).toEqual(
        new ColoredText('pões        ❌', Array(4).fill(Color.Y).concat(Array(9).fill(Color.W)))
      );
    });

    it('renders whole table without past', () => {
      const verb = wordDatabase.verb('conhecer');
      const results = generateResults(verb);
      const printer = new PTVerbConjugation(verb, results);

      const table = printer.getTable();
      const output = new Output();
      output.moveToColoredRows(0, 0, table);

      expect('\n'.concat(output.getOutput())).toMatchSnapshot();
      expect('\n'.concat(output.getColorOutput())).toMatchSnapshot();
    });

    it('renders whole table with past', () => {
      const verb = wordDatabase.verb('falar');
      const results = generateResults(verb);
      const printer = new PTVerbConjugation(verb, results);

      const table = printer.getTable();
      const output = new Output();
      output.moveToColoredRows(0, 0, table);

      expect('\n'.concat(output.getOutput())).toMatchSnapshot();
      expect('\n'.concat(output.getColorOutput())).toMatchSnapshot();
    });
  });

  describe('German Verb', () => {
    it('renders standard', () => {
      const verb = readAllDE().verbs[1];
      const results = generateResultsDE(verb);
      const printer = new DEVerbConjugation(verb, results);

      const exception = printer.renderCell(1, 0);
      const nonException = printer.renderCell(0, 0);

      expect(nonException).toEqual(
        new ColoredText('habe        ✅', Array(3).fill(Color.W).concat(Color.G).concat(Array(9).fill(Color.W)))
      );
      expect(exception).toEqual(
        new ColoredText('hast        ❌', Array(4).fill(Color.Y).concat(Array(9).fill(Color.W)))
      );
    });

    it('renders whole table', () => {
      const verb = readAllDE().verbs[1];
      const results = generateResultsDE(verb);
      const printer = new DEVerbConjugation(verb, results);

      const table = printer.getTable();
      const output = new Output();
      output.moveToColoredRows(0, 0, table);

      expect('\n'.concat(output.getOutput())).toMatchSnapshot();
      expect('\n'.concat(output.getColorOutput())).toMatchSnapshot();
    });
  });

  describe('German Case', () => {
    it('renders case', () => {
      const caseWord = readAllDE().case[0];
      const printer = new DECaseConjugation(caseWord, []);

      const exception = printer.renderCell(1, 0);
      const nonException = printer.renderCell(0, 0);

      expect(nonException).toEqual(
        new ColoredText('habe        ✅', Array(3).fill(Color.W).concat(Color.G).concat(Array(9).fill(Color.W)))
      );
      expect(exception).toEqual(
        new ColoredText('hast        ❌', Array(4).fill(Color.Y).concat(Array(9).fill(Color.W)))
      );
    });

    it('renders whole table', () => {
      const caseWord = readAllDE().case.find((casWord) => casWord.german.nominative === 'ich')!;
      const printer = new DECaseConjugation(caseWord, [
        ...generateResultForExercise(GermanCaseExercise.new(caseWord, GermanCase.nominative), true, 'keyboard', 1)
      ]);

      const table = printer.getTable();
      const output = new Output();
      output.moveToColoredRows(0, 0, table);

      expect('\n'.concat(output.getOutput())).toMatchSnapshot();
      expect('\n'.concat(output.getColorOutput())).toMatchSnapshot();
    });
  });
});
