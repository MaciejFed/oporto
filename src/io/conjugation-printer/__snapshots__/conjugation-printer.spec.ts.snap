// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conjugation Printer German Case renders whole table 1`] = `
"
ich +|mich ~|                                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer German Case renders whole table 2`] = `
"
WWWWWWWWWWWWW                                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer German Verb renders whole table 1`] = `
"
habe        ðŸŸ¢|             |                                         
hast        ðŸ”´|             |                                         
hat         ðŸ”µ|             |                                         
haben       ðŸ”µ|             |                                         
habt        ðŸ”µ|             |                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer German Verb renders whole table 2`] = `
"
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer Portuguese Verb renders whole table with past 1`] = `
"
falo        +|falei       ~|                                          
falas       -|falaste     ~|                                          
fala        ~|falou       ~|                                          
falamos     ~|falÃ¡mos     ~|                                          
falam       ~|falaram     ~|                                          
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer Portuguese Verb renders whole table with past 2`] = `
"
WWWWWWWWWWWWWWWWWWWWWWWWWWWW                                          
WWWWWWWWWWWWWWWWWWWWWWWWWWWW                                          
WWWWWWWWWWWWWWWWWWWWWWWWWWWW                                          
WWWWWWWWWWWWWWWWWWWWWWWWWWWW                                          
WWWWWWWWWWWWWWWWWWWWWWWWWWWW                                          
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer Portuguese Verb renders whole table without past 1`] = `
"
Eu:            conheÃ§o   +| |                                         
Tu:            conheces  -| |                                         
ElaEleVocÃª:    conhece   ~| |                                         
NÃ³s:           conhecemos~| |                                         
ElesElasVosÄ“s: conhecem  ~| |                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Conjugation Printer Portuguese Verb renders whole table without past 2`] = `
"
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
WWWWWWWWWWWWWWWWWWWWWWWWWWWWW                                         
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;
